File: src/main/java/ba/aljovic/amer/movierecommendation/application/algorithm/Recommendation.java
package ba.aljovic.amer.movierecommendation.application.algorithm;

import ba.aljovic.amer.movierecommendation.application.model.UserRating;
import ba.aljovic.amer.movierecommendation.application.utils.Utils;
import org.apache.commons.lang.ArrayUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.mllib.clustering.KMeans;
import org.apache.spark.mllib.clustering.KMeansModel;
import org.apache.spark.mllib.linalg.Vector;
import org.apache.spark.mllib.linalg.Vectors;
import org.apache.spark.mllib.regression.LabeledPoint;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.LinkedList;
import java.util.List;

public class Recommendation
{
    private final RecommendationAlgorithm algorithm;
    private final Integer numberOfClusters;
    private final Integer numberOfIterations;
    private final Double testDataRatio;
    private JavaSparkContext sparkContext;

    public Recommendation(final String master, final RecommendationAlgorithm algorithm, Integer numberOfClusters,
                          Integer numberOfIterations, Double testDataRatio)
    {
        this.algorithm = algorithm;
        this.numberOfClusters = numberOfClusters;
        this.numberOfIterations = numberOfIterations;
        this.testDataRatio = testDataRatio;

        SparkConf configuration = new SparkConf().setAppName(algorithm.getName()).setMaster(master);
        this.sparkContext = new JavaSparkContext(configuration);
    }

    public Double evaluateRecommendationAlgorithm(String data)
    {
        // Read files to RDD
        JavaRDD<UserRating>[] splits = readUserRating(data).randomSplit(new double[]{1 - testDataRatio, testDataRatio});
        JavaRDD<UserRating> trainingData = splits[0];
        JavaRDD<UserRating> testData = splits[1];

        // Train KMeans model
        final KMeansModel ratingClusters = getKMeansModel(trainingData);

        return algorithm.evaluate(trainingData, testData, ratingClusters);
    }

    private KMeansModel getKMeansModel(JavaRDD<UserRating> trainingData)
    {
        JavaRDD<Vector> trainingRatings = trainingData
                .map(UserRating::getData)
                .map(mapToVector());
        return KMeans.train(trainingRatings.rdd(), numberOfClusters, numberOfIterations);
    }


    public Double evaluateManyUsers(final String trainingFolderName) throws IOException
    {
        final String trainingFolder = getClass().getClassLoader().getResource(trainingFolderName).getFile();
        return Files.walk(Paths.get(trainingFolder))
                .filter(filePath -> filePath.toAbsolutePath().toFile().isFile())
                .mapToDouble(filePath -> {
                    String username = filePath.getFileName().toString();
                    return evaluateRecommendationAlgorithm(trainingFolderName + "/" + username);
                })
                .average()
                .getAsDouble();
    }

    private JavaRDD<UserRating> readUserRating(String fileName)
    {
        File trainingDataFile = new File(String.valueOf(getClass().getClassLoader().getResource(fileName)));
        return sparkContext.textFile(trainingDataFile.getPath()).map(mapToUserRating());
    }

    private Function<String, UserRating> mapToUserRating()
    {
        return line -> {
            String[] attributes = line.split("::");
            Integer rating = Integer.valueOf(attributes[2]);
            String[] genomes = attributes[3].split(":");
            LabeledPoint data = new LabeledPoint(rating, Vectors.sparse(genomes.length, range(0, genomes.length - 1),
                    Utils.toDouble(genomes)));
            return new UserRating(rating, data);
        };
    }

    private Function<LabeledPoint, Vector> mapToVector()
    {
        return data -> Vectors.dense(data.label());
    }

    public static int[] range(int min, int max) {
        List<Integer> list = new LinkedList<>();
        for (int i = min; i <= max; i++) {
            list.add(i);
        }

        Integer[] array = list.toArray(new Integer[list.size()]);
        return ArrayUtils.toPrimitive(array);
    }

    //region BUILDER
    public static class RecommendationBuilder
    {
        private RecommendationAlgorithm algorithm;
        private String master = "local[1]";
        private Integer numberOfClusters = 5;
        private Integer numberOfIterations = 20;
        private Double testDataRatio = 0.2;

        public RecommendationBuilder(RecommendationAlgorithm algorithm)
        {
            this.algorithm = algorithm;
        }

        public RecommendationBuilder setMaster(String master)
        {
            this.master = master;
            return this;
        }

        public RecommendationBuilder setTestDataRatio(Double testDataRatio)
        {
            this.testDataRatio = testDataRatio;
            return this;
        }

        public RecommendationBuilder setAlgorithm(RecommendationAlgorithm algorithm)
        {
            this.algorithm = algorithm;
            return this;
        }

        public RecommendationBuilder setNumberOfClusters(Integer numberOfClusters)
        {
            this.numberOfClusters = numberOfClusters;
            return this;
        }

        public RecommendationBuilder setNumberOfIterations(Integer numberOfIterations)
        {
            this.numberOfIterations = numberOfIterations;
            return this;
        }

        public Recommendation build()
        {
            return new Recommendation(master, algorithm, numberOfClusters, numberOfIterations, testDataRatio);
        }
    }
    //endregion
}

File: src/main/java/ba/aljovic/amer/movierecommendation/application/algorithm/Recommendation.scala
package ba.aljovic.amer.movierecommendation.application.algorithm

import java.io.File
import java.lang.String._

import ba.aljovic.amer.movierecommendation.application.model.{Movie, UserRating}
import ba.aljovic.amer.movierecommendation.utils.Utils
import org.apache.commons.io.FileUtils
import org.apache.spark.mllib.clustering.{KMeans, KMeansModel}
import org.apache.spark.mllib.linalg.Vectors._
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

case class Recommendation(algorithm: RecommendationAlgorithm, master: String = "local[1]", numberOfClusters: Integer = 5,
                     numberOfIterations: Integer = 20, testDataRatio: Double = 0.2) extends Serializable
{
  val sparkContext = new SparkContext(new SparkConf().setAppName(algorithm.name()).setMaster(master))

  def evaluateRecommendationAlgorithm(user: String): Double =
  {
    validate(user)
    val splits: Array[RDD[UserRating]] = readUserRatings(user).randomSplit(Array[Double](1 - testDataRatio, testDataRatio))
    val trainingData: RDD[UserRating] = splits(0)
    val testData: RDD[UserRating] = splits(1)
    val ratingClusters: KMeansModel = getKMeansModel(trainingData)
    algorithm.evaluate(trainingData, testData, ratingClusters)
  }

  def recommendMovies(userData: String, moviesData: String): Array[Movie] =
  {
    val userRatings: RDD[UserRating] = readUserRatings(userData)
    val movies: RDD[Movie] = readMovies(moviesData)
    algorithm.recommendMovies(userRatings, movies)
  }

  def evaluateManyUsers(trainingFolderName: String): Double =
  {
    val usersDirectory = new File(Utils.loadFilePath(trainingFolderName))
    val users = FileUtils.listFiles(usersDirectory, null, false).iterator()

    var countUsers = 0
    var result = 0D
    while (users.hasNext)
    {
      val user = users.next()
      val username = Utils.getLastPartFromPath(user)
      try
      {
        result += evaluateRecommendationAlgorithm(trainingFolderName + "/" + username)
      }
      catch
        {
          case e: NumberFormatException => println(e.getMessage)
        }
      countUsers += 1
    }
    result / countUsers
  }



  private def readMovies(fileName: String): RDD[Movie] =
  {
    val moviesFile = new File(valueOf(getClass.getClassLoader.getResource(fileName)))
    val data: RDD[String] = sparkContext.textFile(moviesFile.getPath)

    data.map(line => {
      val attributes = line.split("::")
      val movieId = attributes(0).toInt
      val movieName = attributes(1)
      val genomes = attributes(2).split(":")
      val vector = sparse(
        genomes.length,
        genomes.indices.toArray,
        genomes.toList.map(g => g.toDouble).toArray
      )
      val data = new LabeledPoint(movieId, vector)
      new Movie(movieId, movieName, data)
    })
  }

  private def readUserRatings(fileName: String): RDD[UserRating] =
  {
    val trainingDataFile = new File(Utils.loadFilePath(fileName))
    val data: RDD[String] = sparkContext.textFile(trainingDataFile.getPath)

    data.map(line => {
      try
      {
        val attributes = line.split("::")
        val rating = attributes(2).toInt
        val genomes = attributes(3).split(":")
        val vector = sparse(
          genomes.length,
          genomes.indices.toArray,
          genomes.toList.map(g => g.toDouble).toArray
        )
        val data = new LabeledPoint(rating, vector)
        new UserRating(rating, data)
      }
      catch
        {
          case e: Exception => println(e.getMessage); new UserRating(9, new LabeledPoint(1D, null))
        }
    })
  }

  private def getKMeansModel(trainingData: RDD[UserRating]): KMeansModel =
  {
    val trainingRatings = trainingData.map(ur=>
    {
      dense(ur.data.label)
    })
    KMeans.train(trainingRatings, numberOfClusters, numberOfIterations)
  }

  private def validate(username: String): Unit =
  {
    if (username.contains(":")) throw new NumberFormatException("For input string: ':")
  }
}
